File: lib_common/src/main/java/com/example/common/widget/popupwindow/BasePopupWindow.java
package com.example.common.widget.popupwindow;

import android.app.Activity;
import android.graphics.drawable.ColorDrawable;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.PopupWindow;

import com.example.common.R;

import java.lang.ref.WeakReference;

/**
 * Created by WangYanBin on 2020/7/13.
 * 所有弹框的基类
 */
public class BasePopupWindow extends PopupWindow {
    private boolean dark;
    private WeakReference<Activity> weakActivity;
    private WindowManager.LayoutParams layoutParams;

    public BasePopupWindow(Activity activity) {
        init(activity, false);
    }

    public BasePopupWindow(Activity activity, boolean dark) {
        init(activity, dark);
    }

    private void init(Activity activity, boolean dark) {
        this.dark = dark;
        this.weakActivity = new WeakReference<>(activity);
        this.layoutParams = weakActivity.get().getWindow().getAttributes();
    }

    protected void setPopupWindowContentView(View view) {
        setContentView(view);
        setFocusable(true);
        setOutsideTouchable(true);
        setBackgroundDrawable(new ColorDrawable(android.graphics.Color.TRANSPARENT));
        setAnimationStyle(R.style.pushBottomAnimStyle);//默认底部弹出，可重写
        setHeight(ViewGroup.LayoutParams.WRAP_CONTENT);
        setWidth(ViewGroup.LayoutParams.MATCH_PARENT);
        setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
        setHideAttributes();
    }

    private void setShowAttributes() {
        if (dark) {
            layoutParams.alpha = 0.7f;
            weakActivity.get().getWindow().setAttributes(layoutParams);
        }
    }

    private void setHideAttributes() {
        if (dark) {
            setOnDismissListener(() -> {
                layoutParams.alpha = 1f;
                weakActivity.get().getWindow().setAttributes(layoutParams);
            });
        }
    }

    @Override
    public void showAsDropDown(View anchor) {
        setShowAttributes();
        super.showAsDropDown(anchor);
    }

    @Override
    public void showAsDropDown(View anchor, int xoff, int yoff) {
        setShowAttributes();
        super.showAsDropDown(anchor, xoff, yoff);
    }

    @Override
    public void showAsDropDown(View anchor, int xoff, int yoff, int gravity) {
        setShowAttributes();
        super.showAsDropDown(anchor, xoff, yoff, gravity);
    }

    @Override
    public void showAtLocation(View parent, int gravity, int x, int y) {
        setShowAttributes();
        super.showAtLocation(parent, gravity, x, y);
    }

    public Activity getWeakActivity() {
        return weakActivity.get();
    }

}

File: lib_common/src/main/java/com/dataqin/common/base/BasePopupWindow.kt
package com.dataqin.common.base

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.transition.Slide
import android.transition.Visibility
import android.view.*
import android.widget.PopupWindow
import androidx.viewbinding.ViewBinding
import com.dataqin.common.R
import java.lang.ref.WeakReference
import java.lang.reflect.ParameterizedType

/**
 * Created by WangYanBin on 2020/7/13.
 * 所有弹出窗口的基类，弹框本身操作并不应该复杂，只拿取对应的binding即可
 * 如果具有复杂的交互则直接写成activity加入对应的透明动效
 */
@SuppressLint("NewApi")
abstract class BasePopupWindow<VB : ViewBinding> : PopupWindow {
    protected lateinit var binding: VB
    private var weakActivity: WeakReference<Activity>? = null
    private var layoutParams: WindowManager.LayoutParams? = null
    private var dark = false

    // <editor-fold defaultstate="collapsed" desc="基类方法">
    constructor(activity: Activity) {
        init(activity)
    }

    constructor(activity: Activity, dark: Boolean = false) {
        init(activity, dark)
    }

    private fun init(activity: Activity, dark: Boolean = false) {
        this.weakActivity = WeakReference(activity)
        this.layoutParams = weakActivity?.get()?.window?.attributes
        this.dark = dark
    }

    protected open fun initialize() {
        val type = javaClass.genericSuperclass
        if (type is ParameterizedType) {
            try {
                val vbClass = type.actualTypeArguments[0] as? Class<VB>
                val method = vbClass?.getMethod("inflate", LayoutInflater::class.java)
                binding = method?.invoke(null, weakActivity?.get()?.layoutInflater) as VB
            } catch (ignored: Exception) {
            }
            contentView = binding.root
            isFocusable = true
            isOutsideTouchable = true
            setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            setTransition()
            height = ViewGroup.LayoutParams.WRAP_CONTENT
            width = ViewGroup.LayoutParams.MATCH_PARENT
            softInputMode = WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE
            setDismissAttributes()
        }
    }

    //默认底部弹出，可重写
    protected fun setTransition(setting: Boolean = true) {
        if (setting) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                enterTransition = Slide().apply {
                    duration = 500
                    mode = Visibility.MODE_IN
                    slideEdge = Gravity.BOTTOM
                }
                setExitTransition(Slide().apply {
                    duration = 500
                    mode = Visibility.MODE_OUT
                    slideEdge = Gravity.BOTTOM
                })
            } else {
                animationStyle = R.style.pushBottomAnimStyle
            }
        } else {
//            enterTransition = null
//            exitTransition = null
            animationStyle = -1
        }
    }

    protected fun getActivity() = weakActivity?.get()
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="重写方法">
    override fun showAsDropDown(anchor: View?) {
        setShowAttributes()
        super.showAsDropDown(anchor)
    }

    override fun showAsDropDown(anchor: View?, xoff: Int, yoff: Int) {
        setShowAttributes()
        super.showAsDropDown(anchor, xoff, yoff)
    }

    override fun showAsDropDown(anchor: View?, xoff: Int, yoff: Int, gravity: Int) {
        setShowAttributes()
        super.showAsDropDown(anchor, xoff, yoff, gravity)
    }

    override fun showAtLocation(parent: View?, gravity: Int, x: Int, y: Int) {
        setShowAttributes()
        super.showAtLocation(parent, gravity, x, y)
    }

    private fun setShowAttributes() {
        if (dark) {
            layoutParams?.alpha = 0.7f
            weakActivity?.get()?.window?.attributes = layoutParams
        }
    }

    private fun setDismissAttributes() {
        if (dark) {
            setOnDismissListener {
                layoutParams?.alpha = 1f
                weakActivity?.get()?.window?.attributes = layoutParams
            }
        }
    }
    // </editor-fold>

}